src/stream/src/executor/source/source_backfill_executor.rs

```diff
@@ -641,13 +644,19 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
     /// Otherwise if we break early, but after rescheduling, an unfinished split is migrated to
     /// this actor, we still need to backfill it.
     async fn backfill_finished(&self, states: &BackfillStates) -> StreamExecutorResult<bool> {
+        let persisted_states = self.backfill_state_store.scan().await?;
+        let actor_id = self.actor_ctx.id;
+        tracing::debug!(
+            actor_id,
+            "checking whether source backfill is finished, persisted_states: {:?}, states: {:?}",
+            persisted_states,
+            states
+        );
         Ok(states
             .values()
             .all(|state| matches!(state, BackfillState::Finished))
-            && self
-                .backfill_state_store
-                .scan()
-                .await?
+            && !persisted_states.is_empty()
+            && persisted_states
```

Comment:
3 parallelism, 2 kafka partitions:

We can see `persisted_states` is from 0 -> 1 -> 2

<img width="1352" alt="image" src="https://github.com/user-attachments/assets/6cfb44bc-9edd-40f0-a896-ca44569f8d01">

Reply:
Wait, so checking `!is_empty()` seems not enough. We need to check `len()==num_splits`? 🤡

Reply:
Or does it goes from 0->1->N? If so, checking `len()>1` might be ok 🤔

Reply:
It turns out to be the case:
- first time 0
- second time 1
- third time N

<img width="1352" alt="image" src="https://github.com/user-attachments/assets/95b7e725-317e-48ce-a518-03c0571d9d56">

Reply:
We will fix this by disallowing online scaling https://github.com/risingwavelabs/risingwave/issues/18300